### DISCUSSION TITLE
Is it possible to use wdio to automate the users actions, as a part of the intended UX of the website?

### DISCUSSION THREAD

**Discussion by NoHara42** [webdriverio/webdriverio#7151] [Q&A] [answered]
Hi,

I'm a student who is working on a walkthrough / feature tour for different web pages. 
Forgive me incase this question isn't well-formed.
I'll do my best to clarify.

I want to be able to have the user press a button, and it will automatically fill out forms, and perform actions for the user - as an integral part of the UX.
Is this possible with WebDriverIO?

The way I understand it, is that WDIO is usually used to build test suites for existing web pages, not to be confused with being integrated as part of the UX for a website.

I see the following issues:
1. Normally when using test suites, a new browser session must be initialised to begin programatic control of the browser. 
(Is there a way to attach to the existing browser/driver that the user is currently using?)
2. If this were possible, would this not be an issue from a fundamental security perspective?
(Would the user need to permit that the browser is being automated?)

Many thanks,
-Ned

**Comment by christian-bromann** [comment, marked as answer]
> Is this possible with WebDriverIO?

Absolutely. WebdriverIO uses the WebDriver protocol that is designed to automate what a user can do: operate on a UI

> 1. Is there a way to attach to the existing browser/driver that the user is currently using?

Yes, see the [`attach`](https://webdriver.io/docs/setuptypes/#attachattachoptions) method of the `webdriverio` package.

> 2\. Would the user need to permit that the browser is being automated?

Yes and no. To be able to attach to a running browser, that browser needs to be started through the WebDriver protocol and needs to have a session ID attached to it. That session id is like a secret key and should not be shared. If you access it you potentially can access an automated session from another person if you have access to the WebDriver backend which would pose a security risk.
